scipy/stats/stats.py

```diff
@@ -408,7 +411,19 @@ def gmean(a, axis=0, dtype=None):
             log_a = np.log(np.asarray(a, dtype=dtype))
     else:
         log_a = np.log(a)
-    return np.exp(log_a.mean(axis=axis))
+
+    if weights:
+        weights = np.array(weights, dtype=dtype, copy=False)
+        elif dtype:
+            # Must change the default dtype allowing array type
+            if isinstance(a, np.ma.MaskedArray):
+                weights = np.ma.asarray(weights, dtype=dtype)
+            else:
+                weights = np.asarray(weights, dtype=dtype)
```

Comment:
I just reconsidered this. The point that if `weights` is already a masked array, you want it to stay a masked array, and if it is not, you want to make it a regular array, right? That would be OK?
If so, how about
```
if weights isinstance(weights, np.ma.MaskedArray):
   weights = weights.astype(dtype)
else:
   weights = np.array(weights, dtype=dtype, copy=False)
```
And consider putting that in a try - except so if the user enters something wonky you can give them a helpful message.

Reply:
If I understood correctly, then that is indeed the intention. However, I just based the input conversion on that of the `a` parameter, because I thought it would be plausible to apply the same methods to both inputs.

Reply:
The suggested changes resulted in broken code, so I reverted 049d1f60634e54db506599481294961092b6918c in cd8ad1bb894343ec6d318e917bb0225b392177a2 for now.

Reply:
> Note that this will also need a unit test.

@rgommers and @mdhaber How many and what kinds of test would you suggest? Do all the existing tests need to be repeated with weights?

Reply:
OK. I would suggest that the four lines above express that intent more directly and on average executes with fewer `if` checks. 

There's at least two other advantage of the code above in special cases:

Suppose `weights` is input as a regular NumPy array with `dtype` `np.float64`, and `log_a` is not a masked array. If the user enters the argument `dtype=np.float64`, then:
`if not isinstance(weights, np.ndarray)` will be false
`if isinstance(a, np.ma.MaskedArray):` will be false
`weights = np.asarray(weights, dtype=dtype)` will run. This will copy `weights` unnecessarily.

Suppose `weights` is input as a regular NumPy array with `dtype` `np.float64`, and `log_a` is a masked array. If the user enters the argument `dtype=np.float64`, then:
`if not isinstance(weights, np.ndarray)` will be false
`if isinstance(a, np.ma.MaskedArray):` will be True
`weights = np.ma.asarray(weights, dtype=dtype)` will run. This converts `weights` to a masked array unnecessarily.

Not blocking, so you can leave as-is if you prefer what you have, but maybe other reviewers will weight in.

Reply:
I trust your expertise and incorporated your changes in 55ff9d2.